src/clients/c++/library/request_grpc.cc

```diff
@@ -1073,12 +1053,27 @@ InferGrpcStreamContextImpl::Run(ResultMap* results)
 {
   // Actually calling AsyncRun() and GetAsyncRunResults()
   std::shared_ptr<Request> req;
-  Error err = AsyncRun(&req);
+  bool callback_invoked = false;
+  std::mutex mtx;
+  std::condition_variable cv;
+  Error err =
+      AsyncRun([&](InferContext* ctx, std::shared_ptr<Request> request) {
+        // Defer the response retrieval to main thread
+        std::lock_guard<std::mutex> lk(mtx);
+        callback_invoked = true;
+        req = std::move(request);
+        cv.notify_all();
+        return;
+      });
   if (!err.IsOk()) {
     return err;
   }
-  bool is_ready;
-  return GetAsyncRunResults(results, &is_ready, req, true);
+  // Wait for the callback
+  {
+    std::unique_lock<std::mutex> lk(mtx);
+    cv.wait(lk, [&callback_invoked]() { return callback_invoked; });
+  }
```

Comment:
The above can be simplified with promise and future if you want.

Reply:
Thanks! Done.